sample_size/sample_size_calculator.py

```diff
@@ -70,46 +50,10 @@ def _get_single_sample_size(self, metric: BaseMetric) -> float:
         return sample_size
 
     def get_sample_size(self) -> float:
-        # Supports the sample size calculation for single metric now.
-        # The current structure is set up to support multiple metrics in the future.
-        sample_size = float("nan")
-
-        if len(self.boolean_metrics) + len(self.numeric_metrics) + len(self.ratio_metrics) == 1 and self.variants == 2:
-            if self.boolean_metrics:
-                sample_size = self._get_single_sample_size(self.boolean_metrics[0])
-            elif self.numeric_metrics:
-                sample_size = self._get_single_sample_size(self.numeric_metrics[0])
-            elif self.ratio_metrics:
-                sample_size = self._get_single_sample_size(self.ratio_metrics[0])
-
+        if len(self.metrics) == 1 and self.variants == DEFAULT_VARIANTS:
+            sample_size = self._get_single_sample_size(self.metrics[0])
         else:
-            min_standardized_effect_size: float = 0
-            if self.boolean_metrics:
-                for bool_metric in self.boolean_metrics:
-                    if min_standardized_effect_size > bool_metric.mde / np.sqrt(bool_metric.variance):
-                        min_standardized_effect_size = bool_metric.mde / np.sqrt(bool_metric.variance)
-                        min_metric = bool_metric
-            if self.numeric_metrics:
-                for numeric_metric in self.numeric_metrics:
-                    if min_standardized_effect_size > numeric_metric.mde / np.sqrt(numeric_metric.variance):
-                        min_standardized_effect_size = numeric_metric.mde / np.sqrt(numeric_metric.variance)
-                        min_metric = numeric_metric
-            if self.ratio_metrics:
-                for ratio_metric in self.ratio_metrics:
-                    if min_standardized_effect_size > ratio_metric.mde / np.sqrt(ratio_metric.variance):
-                        min_standardized_effect_size = ratio_metric.mde / np.sqrt(ratio_metric.variance)
-                        min_metric = ratio_metric
-            power_analysis = min_metric.default_power_analysis_instance
-            sample_size = get_multiple_sample_size(min_standardized_effect_size, power_analysis)
+            sample_size = MultipleTesting(self.metrics, self.variants, self.alpha,
```

Comment:
All 4 of the arguments you're passing to `MultipleTesting` are from the state of the `SampleSizeCalculator` instance. I think encapsulating the logic around mutliple testing into its own class is fine, but a better way of doing that might be the use of a "mix-in" class

```python
from .multiple_testing_funcs import MultipleTestingMixin

class SampleSizeCalculator(MultipleTestingMixin):
     def get_sample_size(self) -> float:
        return self.get_multiple_sample_size()

class MultipleTestingMixin:
    # note no __init__
    def get_multiple_sample_size(self) -> int:
        if len(self.metrics) < 2:
            return self.metrics[0].single_sample_size()
        lower = max([metric.single_sample_size(self.alpha, self.power) for metric in self.metrics])
        upper = max([metric.single_sample_size(self.alpha / self.m, self.power) for metric in self.metrics])

        return self._find_sample_size(lower, upper)

    def _find_sample_size(self, lower: float, upper: float, depth=0):
        MAX_RECURSION_DEPTH = 20
        EPSILON = 0.025

        if depth > MAX_RECURSION_DEPTH:
            raise RecursionError

        candidate = int((upper + lower) / 2)
        expected_power = self._expected_average_power(candidate)

        if np.isclose((self.power, expected_power), EPSILON) or np.isclose((candidate, upper), EPSILON):
            return candidate

        if expected_power > self.power:
            return self._find_sample_size(lower, candidate, depth + 1)
        else:
            return self._find_sample_size(candidate, upper, depth + 1)

    def _expected_average_power(self, sample_size: int):
        power = []
        for m1 in range(1, self.m + 1):
            alts = np.array([True] * m1 + [False] * (self.m - m1))
            for _ in range(self.REPLICATION):
                true_alt = alts[np.random.permutation(self.m)]
                p_values = [self._generate_p_value(m, true_alt[i], sample_size) for i, m in enumerate(self.metrics)]
                rejected = multipletests(p_values, alpha=self.alpha, method="fdr_bh")[0]
                power.append(np.dot(rejected, true_alt) / m1)

        return np.mean(power)
```